core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+   * performed automatically when the set is modified, but some operations (like
+   * e.g. `size`) do not modify the set and therefore do not trigger cleanup.
+   */
+  def gc(): Unit =
+    this.segments.foreach(segment => segment.restructureIfNecessary(false))
+
+  /**
+   * Get estimated number of elements in the set. This method does not perform
+   * cleanup of dead references, so it might return a little bit higher value
+   * than the actual number of elements.
+   *
+   * @return
+   *   estimated number of elements in the set
+   */
+  override def size(): Int =
+    this.segments.map(segment => segment.size()).sum
+
+  /**
+   * Check if the set is empty.
+   *
+   * @return
+   *   `true` if the set is empty, `false` otherwise
+   * @see
+   *   [[ConcurrentWeakHashSet.size]]
+   */
+  override def isEmpty: Boolean =
+    this.segments.forall(segment => segment.size() == 0)
+
+  /**
+   * @return
+   *   specialized iterator that excludes dead references
+   */
+  override def iterator: Iterator[V] =
+    new ConcurrentWeakHashSetIterator()
+
+  /**
+   * Check if the set contains the specified element.
+   *
+   * @param element
+   *   element to check
+   * @return
+   *   `true` if the set contains the specified element, `false` otherwise
+   */
+  override def contains(element: V): Boolean = {
+    if (element == null) false
+    val reference     = getReference(element)
+    val storedElement = if (element != null) reference.get() else null
+    storedElement != null
+  }
+
+  /**
+   * Add given elements to the set.
+   *
+   * @param element
+   *   elements to add
+   * @return
+   *   `true` if the element was added, `false` otherwise
+   * @see
+   *   [[ConcurrentWeakHashSet.add]]
+   */
+  def addAll(elements: Iterable[V]): Unit =
+    elements.foreach(this.add)
+
+  /**
+   * Add given element to the set.
+   *
+   * @param element
+   *   element to add
+   * @return
+   *   instance of the set
+   */
+  override def addOne(element: V): this.type = {
+    this.add(element)
+    this
+  }
+
+  /**
+   * Add given element to the set. Remember that this method does not support
+   * `null` elements, because it's not a valid value for `WeakReference`.
+   *
+   * @param element
+   *   not-null element to add
+   * @return
+   *   `true` if the element was added, `false` otherwise
+   * @throws IllegalArgumentException
+   *   if the element is `null`
+   */
+  override def add(element: V): Boolean = {
+    if (element == null) throw new IllegalArgumentException("ConcurrentWeakHashSet does not support null elements.")
+    this.update(
+      element,
+      new ConcurrentWeakHashSet.UpdateOperation[V](
+        ConcurrentWeakHashSet.UpdateOptions.RestructureBefore,
+        ConcurrentWeakHashSet.UpdateOptions.Resize
+      ) {
+        override def execute(oldRef: Ref[V]): ConcurrentWeakHashSet.UpdateResults.UpdateResult =
+          if (oldRef == null || oldRef.get() != element)
```

Comment:
Consider the trick `(oldRef.asInstanceOf[AnyRef] eq null)` here and anywhere else you are comparing with null.

Also for a reference check, `eq` is faster than `==` or `!=`. So I think you want to use `.ne` here (standing for "not equal"). If Scala complains it's only because it doesn't know `oldRef.get()` is a reference type, so you can just cast to `AnyRef`.

In general, in high-performance code, `!=` and `==` should be avoided as they go to the slow `.equals` method of `java.lang.Object`, which generally does structural equality.

Reply:
Are sure about the performance of `==` vs `eq` in terms of comparing it to the null? It looks like both scenarios are optimized by Scala compiler and none of them call the `equals` in such scenario:

* https://scastie.scala-lang.org/Nvw3tVfMSjmEd6uWKgHssA

I'm just curious, probably there's something I've missed 😅